File: sdks/python/conformance/run_conformance.py
#!/usr/bin/env python3
"""CLI entry point: run the 65 contracts fixtures, emit the compatibility
report (contracts/README.md schema — fixtureId/suite/language/status/
limitation/message rows, same shape node/go/java write).

Usage::

    python conformance/run_conformance.py [--contracts PATH] [--out PATH]

Exit code is non-zero when any fixture fails (harness.md runner obligation).
"""

from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))

from runner import run_all  # noqa: E402


def main() -> int:
    repo_root = Path(__file__).resolve().parents[3]
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--contracts", type=Path, default=repo_root / "contracts",
        help="Path to the contracts fixture directory",
    )
    parser.add_argument(
        "--out", type=Path,
        default=Path(__file__).resolve().parent / "compatibility-report.python.json",
        help="Where to write the compatibility report JSON",
    )
    args = parser.parse_args()

    report = run_all(args.contracts)
    args.out.write_text(json.dumps(report, indent=2, default=str) + "\n")

    summary = report["summary"]
    print(
        f"conformance[python]: {summary['pass']} passed, {summary['fail']} failed, "
        f"{summary['skipped-with-documented-limitation']} skipped-with-documented-limitation, "
        f"{summary['skipped-v1-out-of-scope']} skipped-v1-out-of-scope "
        f"(report: {args.out})"
    )
    for row in report["results"]:
        if row["status"] == "fail":
            print(f"  FAIL {row['suite']}/{row['fixtureId']}")
            if row["message"]:
                print(f"       - {row['message']}")
        elif row["status"] not in ("pass",):
            print(f"  SKIP {row['suite']}/{row['fixtureId']}: {row['limitation']}")
    return 1 if summary["fail"] else 0


if __name__ == "__main__":
    raise SystemExit(main())

File: sdks/dart/conformance/run_conformance.dart
import 'dart:convert';
import 'dart:io';

import 'runner.dart';

/// CLI entry point: run the `contracts/` fixtures this SDK's architecture
/// can represent and emit the compatibility report (`contracts/README.md`
/// schema — the same shape node/python/go/java/rust/swift write).
///
/// Usage (from `sdks/dart`):
///
///     dart run conformance/run_conformance.dart \
///       --contracts ../../contracts \
///       --out conformance/compatibility-report.dart.json
///
/// Exit code is non-zero when any fixture fails (`contracts/harness.md`
/// runner obligation 6).
Future<void> main(List<String> args) async {
  var contracts = '../../contracts';
  var out = 'conformance/compatibility-report.dart.json';
  for (var i = 0; i < args.length; i += 1) {
    switch (args[i]) {
      case '--contracts':
        i += 1;
        if (i >= args.length) {
          _die('--contracts requires a path');
        }
        contracts = args[i];
      case '--out':
        i += 1;
        if (i >= args.length) {
          _die('--out requires a path');
        }
        out = args[i];
      default:
        _die('unknown argument ${args[i]}');
    }
  }

  final contractsDir = Directory(contracts);
  if (!contractsDir.existsSync()) {
    _die('contracts directory not found: ${contractsDir.absolute.path}');
  }

  final report = await runAll(contractsDir);

  final outFile = File(out);
  outFile.parent.createSync(recursive: true);
  const encoder = JsonEncoder.withIndent('  ');
  outFile.writeAsStringSync('${encoder.convert(_sorted(report.toJson()))}\n');

  final summary = report.summary();
  stdout.writeln(
    'conformance[$language]: ${summary[Status.pass]} passed, '
    '${summary[Status.fail]} failed, '
    '${summary[Status.skippedWithDocumentedLimitation]} '
    'skipped-with-documented-limitation, '
    '${summary[Status.skippedV1OutOfScope]} skipped-v1-out-of-scope '
    '(report: $out)',
  );

  if ((summary[Status.fail] ?? 0) > 0) {
    for (final row in report.results.where((r) => r.status == Status.fail)) {
      stdout.writeln(
        '  FAIL ${row.suite}/${row.fixtureId}'
        '${row.message == null ? '' : ' - ${row.message}'}',
      );
    }
    exit(1);
  }
}

Never _die(String message) {
  stderr.writeln('run_conformance: $message');
  exit(2);
}

/// Sorted keys at every level, for a byte-stable report (mirrors swift's
/// `.sortedKeys` and go/python's canonical JSON).
Object? _sorted(Object? value) {
  if (value is Map) {
    final keys = value.keys.map((k) => k.toString()).toList()..sort();
    return <String, Object?>{for (final key in keys) key: _sorted(value[key])};
  }
  if (value is List) {
    return value.map(_sorted).toList();
  }
  return value;
}
